apps/client/src/common/stores/appModeStore.ts

```diff
@@ -5,66 +5,146 @@ export enum AppMode {
 }
 
-async function persistModeToSession(mode: AppMode) {
+function persistModeToSession(mode: AppMode) {
   localStorage.setItem(appModeKey, mode);
 }
 
+export type EventToEdit = { id: string; index: number; anchor: boolean };
+
 type AppModeStore = {
   mode: AppMode;
   cursor: string | null;
-  editId: string | null;
+  editMode: EditMode;
+  eventsToEdit: EventToEdit[];
+  isEventSelected: (id: string, index: number, isIndividual?: boolean) => boolean;
   setMode: (mode: AppMode) => void;
-  setCursor: (id: string | null, isEditable?: boolean) => void;
-  setEditId: (id: string | null) => void;
+  setEditMode: (mode: EditMode) => void;
+  setEventsToEdit: (id: string, index: number) => void;
+  clearEventsToEdit: () => void;
 };
 
-export const useAppMode = create<AppModeStore>()((set) => ({
+export const useAppMode = create<AppModeStore>()((set, get) => ({
   mode: getModeFromSession(),
   cursor: null,
-  editId: null,
-  setMode: (mode: AppMode) =>
-    set((state) => {
-      persistModeToSession(mode);
-      return mode === AppMode.Edit
-        ? {
-            editId: state.cursor,
-            mode: mode,
-          }
-        : {
-            editId: null,
-            mode: mode,
-          };
-    }),
-  setCursor: (id: string | null, isEditable?: boolean) =>
-    set((state) => {
-      if (isEditable) {
-        return state.mode === AppMode.Edit
-          ? {
-              cursor: id,
-              editId: id,
+  editMode: EditMode.Single,
+  eventsToEdit: [],
+  setMode: (mode: AppMode) => {
+    persistModeToSession(mode);
+
+    return set(() => {
+      return { mode };
+    });
+  },
+  isEventSelected: (id, index, isIndividual = false) => {
+    const { eventsToEdit } = get();
+
+    if (isIndividual && eventsToEdit.length > 1) {
+      return false;
+    }
+
+    return eventsToEdit.some((event) => {
+      const isAnchorPresent = eventsToEdit.some((event) => event.anchor);
+
+      if (isAnchorPresent) {
+        const firstSelectedEvent = eventsToEdit.at(0);
+        const lastSelectedEvent = eventsToEdit.at(-1);
+
+        if (!firstSelectedEvent || !lastSelectedEvent) {
+          return event.id === id;
+        }
+
+        if (firstSelectedEvent.index <= index && lastSelectedEvent.index >= index) {
+          return true;
+        }
+      }
+
+      return event.id === id;
+    });
+  },
+  setEditMode: (mode) => set(() => ({ editMode: mode })),
+  setEventsToEdit: (id, index) =>
+    set(({ editMode, eventsToEdit }) => {
+      if (editMode === EditMode.Single) {
+        return { eventsToEdit: [{ id, index, anchor: false }] };
+      }
+
+      if (editMode === EditMode.CherryPick) {
+        const uniqueEventsWithoutAnchor = eventsToEdit
+          .map(({ id, index }) => ({ id, index, anchor: false }))
+          // in case the user toggles the same event,
+          // we can preemptively filter out duplicates
+          .filter((event) => event.id !== id);
```

Comment:
nice, I reckon it can potentially be more efficient by filtering first

Reply:
Nice catch!